SwiftCompilerSources/Sources/Optimizer/ModulePasses/DiagnoseUnknownConstValues.swift

```diff
@@ -0,0 +1,128 @@
+
+  init(bridged: BridgedConstExprFunctionState) { self.bridged = bridged }
+  init() {
+    self.bridged = BridgedConstExprFunctionState.create()
+  }
+  
+  mutating func isConstantValue(_ value: Value) -> Bool {
+    return bridged.isConstantValue(value.bridged)
+  }
+  
+  /// TODO: once we have move-only types, make this a real deinit.
+  mutating func deinitialize() {
+    bridged.deinitialize()
+  }
+}
+
+/// Performs mandatory diagnostic pass for emitting errors for '@const' values
+/// which the compiler was not able to simplify/interpret/reduce to a symbolic
+/// value.
+///
+let diagnoseUnknownConstValues = ModulePass(name: "diagnose-unknown-const-values") {
+  (moduleContext: ModulePassContext) in
+  var constExprState: ConstExprFunctionState = .init()
+  defer { constExprState.deinitialize() }
+  
+  // Verify all const globals to be initialized with
+  // compile-time known values
+  verifyGlobals(moduleContext);
+  
+  // Verify @const lets appearing as local variables
+  verifyLocals(moduleContext,
+               constExprState: &constExprState);
+  
+  // For each function call, ensure arguments to @const parameters
+  // are all compile-time known values
+  verifyCallArguments(moduleContext,
+                      constExprState: &constExprState);
+  
+  // For each `@const` function, ensure it is fully
+  // evaluable/interpretable at compile time
+  verifyFunctions(moduleContext);
+}
+
+private func verifyGlobals(_ context: ModulePassContext) {
+  for gv in context.globalVariables.filter({ $0.isConst }) {
+    if gv.staticInitValue == nil {
+      context.diagnosticEngine.diagnose(gv.varDecl!.location.sourceLoc,
+                                        .require_const_initializer_for_const)
+    }
+  }
+}
+
+private func verifyLocals(_ context: ModulePassContext,
+                          constExprState: inout ConstExprFunctionState) {
+  for f in context.functions {
+    for bb in f.blocks {
+      for i in bb.instructions {
+        if let dbi = i as? DebugValueInst {
+          verifyLocal(debugValueInst: dbi, in: context,
+                      constExprState: &constExprState)
+        }
+      }
+    }
+  }
+}
+
+private func verifyLocal(debugValueInst: DebugValueInst, in context: ModulePassContext,
+                         constExprState: inout ConstExprFunctionState) {
+  guard let localVarDecl = debugValueInst.varDecl,
+        !localVarDecl.isParamDecl,
+        localVarDecl.isConst else {
+    return
+  }
+  
+  if !constExprState.isConstantValue(debugValueInst.operand.value) {
+    context.diagnosticEngine.diagnose(debugValueInst.location.sourceLoc,
+                                      .require_const_initializer_for_const)
+  }
+}
+
+private func verifyCallArguments(_ context: ModulePassContext,
+                                 constExprState: inout ConstExprFunctionState) {
+  for f in context.functions {
+    for bb in f.blocks {
+      for i in bb.instructions {
+        if let apply = i as? ApplySite {
```

Comment:
Do you want to do this check also for `partial_apply` instruction? I.e. for closures?
If not, you want to cast to `FullApplySite` instead